annotated.go

```diff
@@ -89,3 +90,268 @@ func (a Annotated) String() string {
 	}
 	return fmt.Sprintf("fx.Annotated{%v}", strings.Join(fields, ", "))
 }
+
+type annotations struct {
+	ParamTags  []string
+	ResultTags []string
+}
+
+// Annotation can be passed to Annotate(f interface{}, anns ...Annotation)
+// for annotating the parameter and result types of a function.
+type Annotation interface {
+	apply(*annotations) error
+	getAnnotatedType(reflect.Type) []reflect.Type
+}
+
+type annotationError struct {
+	target interface{}
+	err    error
+}
+
+func (e *annotationError) Error() string {
+	return e.err.Error()
+}
+
+type paramTags struct {
```

Comment:
Consider naming this paramTagsAnnotation to make the interface compliance clearer.
Also, optionally, add an interface compliance check.

https://github.com/uber-go/guide/blob/master/style.md#verify-interface-compliance